flowmachine/flowmachine/features/subscriber/mobility_classification.py

```diff
@@ -0,0 +1,103 @@
+from flowmachine.features.utilities.subscriber_locations import BaseLocation
+from flowmachine.features.subscriber.subscriber_stay_lengths import (
+    SubscriberStayLengths,
+)
+from flowmachine.core.errors import InvalidSpatialUnitError
+
+
+class MobilityClassification(SubscriberFeature):
+    """
+    Based on subscribers' reference locations in a sequence of reference
+    periods, classify each subscriber as having one of the following mobility
+    types (the assigned label corresponds to the first of these criteria that
+    is true for a given subscriber):
+
+    - 'unlocated': Subscriber has a NULL location in the most recent period
+    - 'irregular': Subscriber is not active in at least one of the reference
+      periods
+    - 'not_always_locatable': Subscriber has a NULL location in at least one of
+      the reference periods
+    - 'mobile': Subscriber spent fewer than 'stay_length_threshold' consecutive
+      periods at any single location
+    - 'stable': Subscriber spent at least 'stay_length_threshold' consecutive
+      periods at the same location
+
+    Only subscribers appearing in the result of the reference location query
+    for the most recent period are included in the result of this query (i.e.
+    subscribers absent from the query result can be assumed to fall into a
+    sixth category: "not active in the most recent period").
+
+    Parameters
+    ----------
+    locations : list of BaseLocation
+        List of reference location queries, each returning a single location
+        per subscriber (or NULL location for subscribers that are active but
+        unlocatable). The list is assumed to be sorted into ascending
+        chronological order.
+    stay_length_threshold : int, default 3
+        Minimum number of consecutive periods over which a subscriber's
+        location must remain the same for that subscriber to be classified as
+        'stable'.
+    """
+
+    def __init__(self, locations: List[BaseLocation], stay_length_threshold: int = 3):
+        self.locations = locations
+        if len(set(l.spatial_unit for l in self.locations)) > 1:
+            raise InvalidSpatialUnitError(
+                "MobilityClassification requires all input locations to have the same spatial unit"
+            )
+        self.stay_length_threshold = int(stay_length_threshold)
+        self.max_stay_lengths = SubscriberStayLengths(
+            locations=self.locations,
+            statistic="max",
+        )
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["subscriber", "value"]
+
+    def _make_query(self) -> str:
+        loc_cols_string = ", ".join(self.locations[0].spatial_unit.location_id_columns)
+
+        # Note: in some ways it would be nicer to use a DayTrajectories query instead of a list of location queries,
+        # but DayTrajectories only works with queries that have 'start' and 'stop' attributes
```

Comment:
I guess, given the custom name of the column, using a Join object is out of the question here? (I guess it might improve caching, but I don't know how much impact it'd have)

Reply:
A `Union` object wouldn't work here because of the need to add the 'ordinal' column. We could use the same hack as in DayTrajectories, and wrap this up in a CustomQuery object, but I don't really like that hack in the place it's already used so I'm reluctant to use it again here.

There could be a benefit to caching this, since the same union is used in the SubscriberStayLengths query, but it's not guaranteed to help (for the same reason CTEs can lead to bad performance - if this bit of SQL is included in the full query then postgres might be able to do some optimisations that wouldn't be possible if it's pre-calculated and just gets read from disk). And if we do cache the union then we're doubling the disk usage, because the results of the location queries will all be cached separately and then cached again in the unioned query.